main/src/ca/uwaterloo/flix/language/phase/sjvm/BytecodeCompiler.scala

```diff
@@ -17,34 +33,105 @@ object BytecodeCompiler {
   sealed trait F[T]
 
   def compileExp[R <: Stack, T <: JType](exp: Expression[T]): F[R] => F[R ** T] = exp match {
-    case Expression.Unit(loc) => pushUnit[R]()
-    case Expression.Null(tpe, loc) => pushNull[R]()
-    case Expression.True(loc) => pushBool[R](true)
-    case Expression.False(loc) => pushBool[R](false)
-    case Expression.Char(lit, loc) => ???
+    case Expression.Unit(loc) =>
+      WithSource[R](loc) ~ pushUnit()
+
+    case Expression.Null(tpe, loc) =>
+      WithSource[R](loc) ~ pushNull()
+
+    case Expression.True(loc) =>
+      WithSource[R](loc) ~ pushBool(true)
+
+    case Expression.False(loc) =>
+      WithSource[R](loc) ~ pushBool(false)
+
+    case Expression.Char(lit, loc) =>
+      WithSource[R](loc) ~ pushChar(lit)
+
     case Expression.Float32(lit, loc) => ???
+
     case Expression.Float64(lit, loc) => ???
+
     case Expression.Int8(lit, loc) => ???
+
     case Expression.Int16(lit, loc) => ???
+
     case Expression.Int32(lit, loc) => ???
+
     case Expression.Int64(lit, loc) => ???
-    case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) => branch(compileExp(exp1), compileExp(exp2), compileExp(exp3))
+
+    case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+      branch(compileExp(exp1), compileExp(exp2), compileExp(exp3))
+
     case Expression.Ref(exp, tpe, loc) => ???
     //      NEW("class name") ~
     //      DUP ~
     //      compileExp(exp) ~
     //      INVOKESPECIAL("class name", "constructor signature")
+
     case Expression.Let(sym, exp1, exp2, tpe, loc) =>
-//      exp1.tpe match {
-//        case ErasedType.Int32() => WithSource[R](null) ~ compileExp(exp1) ~ popInt()
-//        case _ => ???
-//      }
-      compileExp(exp2)
+      // sym is unsafe. localvar stack + reg as types?
+      WithSource[R](loc) ~
+        compileExp(exp1) ~
+        XStore(sym, exp1.tpe) ~
+        compileExp(exp2)
+
+    case Expression.ArrayLoad(base, index, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        compileExp(index) ~
+        ArrayInstructions.XALoad(tpe)
+
+    case Expression.ArrayStore(base, index, elm, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        compileExp(index) ~
+        compileExp(elm) ~
+        ArrayInstructions.XAStore(elm.tpe) ~
+        pushUnit()
+
+    case Expression.ArrayLength(base, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        ArrayInstructions.arrayLength()
+
+    case Expression.ArraySlice(base, beginIndex, endIndex, tpe, loc) =>
+      WithSource[R](loc) ~
+        compileExp(base) ~
+        compileExp(beginIndex) ~
+        compileExp(endIndex) ~
+        SWAP(beginIndex.tpe, endIndex.tpe) ~
+        DUP_X1(endIndex.tpe, beginIndex.tpe) ~
+        ISUB() ~
+        DUP()
+        // newArray(tpe) ~ not sure how to continue
```

Comment:
While there are only two bytecode instructions for arrays, NEWARRAY and 
ANEWARRAY there are really more than that because these NEWARRAY instructions takes a type descriptor. Thus I think we should just create ANEWARRAY, DNEWARRAY, and so forth.

Reply:
https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-6.html#jvms-6.5.newarray